tsl/src/compression/compression.c

```diff
@@ -1450,3 +1452,58 @@ compression_get_toast_storage(CompressionAlgorithms algorithm)
 		elog(ERROR, "invalid compression algorithm %d", algorithm);
 	return definitions[algorithm].compressed_data_storage;
 }
+
+/* Get relstats from compressed chunk and insert into relstats for the
+ * corresponding uncompressed parent.
+ */
+extern void
+update_compressed_chunk_relstats(Oid uncompressed_relid, Oid compressed_relid)
+{
+	bool update = true;
+	double rowcnt;
+	int comp_pages, uncomp_pages, comp_visible, uncomp_visible;
+	float comp_tuples, uncomp_tuples, out_tuples;
+	Chunk *uncompressed_chunk = ts_chunk_get_by_relid(uncompressed_relid, true);
+	Chunk *compressed_chunk = ts_chunk_get_by_relid(compressed_relid, true);
+
+	if (uncompressed_chunk->table_id != uncompressed_relid ||
+		uncompressed_chunk->fd.compressed_chunk_id != compressed_chunk->fd.id ||
+		compressed_chunk->table_id != compressed_relid)
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_INTERNAL_ERROR),
+				 errmsg("mismatched chunks for relstats update %d %d",
+						uncompressed_relid,
+						compressed_relid)));
+	}
+
+	/* we need page info from compressed relid */
+	capture_pgclass_stats(compressed_relid, &comp_pages, &comp_visible, &comp_tuples);
+	rowcnt = (double) ts_get_before_compression_row_count(uncompressed_chunk->fd.id);
+	capture_pgclass_stats(uncompressed_relid, &uncomp_pages, &uncomp_visible, &uncomp_tuples);
+
+	/* Before compressing a chunk in 2.0, we save its stats. Prior
+	 * releases do not support this. So the stats on uncompressed relid
+	 * could be invalid. In this case, do the best that we can.
+	 * Alternately, the stats could match, don't update in that case.
+	 */
+	if (uncomp_tuples == 0)
+	{
+		if (rowcnt > 0)
+			out_tuples = (float4) rowcnt;
+		else
+			out_tuples = (float4) comp_tuples;
+	}
+	else
+	{
+		out_tuples = (float4) uncomp_tuples;
```

Comment:
We would reach this path only in 2.0 as we set uncomp_tuples to a non zero value. If we delete rows from the compressed adta, it does not change rowcnt, but would modify comp_tuples. The comp_tuples count could be unexpectedly low (based on the compression ratio). I have chosen to over-estimate the output rows rather than under-estimate by retaining the old value for uncomp_tuples..